substrate/frame/message-queue/src/lib.rs

```diff
@@ -1447,7 +1447,22 @@ impl<T: Config> Pallet<T> {
 		use ProcessMessageError::*;
 		let prev_consumed = meter.consumed();
 
-		match T::MessageProcessor::process_message(message, origin.clone(), meter, &mut id) {
+		let transaction =
+			storage::with_transaction(|| -> TransactionOutcome<Result<_, DispatchError>> {
+				let res =
+					T::MessageProcessor::process_message(message, origin.clone(), meter, &mut id);
+				match &res {
+					Ok(_) => TransactionOutcome::Commit(Ok(res)),
+					Err(_) => TransactionOutcome::Rollback(Ok(res)),
+				}
+			});
+
+		let transaction = match transaction {
+			Ok(result) => result,
+			_ => unreachable!(),
```

Comment:
With current `storage::with_transaction` implementation, this is indeed unreachable, but if the internals of that function change, this will crash the runtime.

I know, highly unlikely, but still I would prefer to:
- `debug_assert!()` to have it crash in tests if the invariant is ever broken,
- return some error `MessageExecutionStatus` in prod without crashing.

Reply:
Yea it should be defensive and then just return an error or so. Having panics in the runtime is never good.